kube/src/config/file_config.rs

```diff
@@ -140,15 +142,168 @@ impl Kubeconfig {
+        }
+    }
+
+    /// Merge kubeconfig file according to the rules described in
+    /// <https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#merging-kubeconfig-files>
+    ///
+    /// > Merge the files listed in the `KUBECONFIG` environment variable according to these rules:
+    /// >
+    /// > - Ignore empty filenames.
+    /// > - Produce errors for files with content that cannot be deserialized.
+    /// > - The first file to set a particular value or map key wins.
+    /// > - Never change the value or map key.
+    /// >   Example: Preserve the context of the first file to set `current-context`.
+    /// >   Example: If two files specify a `red-user`, use only values from the first file's `red-user`.
+    /// >            Even if the second file has non-conflicting entries under `red-user`, discard them.
+    fn merge(mut self, mut next: Kubeconfig) -> Result<Self> {
+        if self.kind.is_some() && next.kind.is_some() && self.kind != next.kind {
+            return Err(ConfigError::KindMismatch.into());
+        }
+        if self.api_version.is_some() && next.api_version.is_some() && self.api_version != next.api_version {
+            return Err(ConfigError::ApiVersionMismatch.into());
+        }
+
+        if self.kind.is_none() {
+            self.kind = next.kind.take();
+        }
+        if self.api_version.is_none() {
+            self.api_version = next.api_version.take();
+        }
+
+        match self.preferences.as_mut() {
+            Some(base_prefs) => {
+                if let Some(mut next_prefs) = next.preferences.take() {
+                    if base_prefs.colors.is_none() {
+                        base_prefs.colors = next_prefs.colors.take();
+                    }
+
+                    match base_prefs.extensions.as_mut() {
+                        Some(mut base_exts) => {
+                            if let Some(next_exts) = next_prefs.extensions {
+                                append_new_named(&mut base_exts, next_exts, |x| &x.name);
+                            }
+                        }
+                        None => {
+                            base_prefs.extensions = next_prefs.extensions.take();
+                        }
+                    }
+                }
+            }
+
+            None => {
+                self.preferences = next.preferences.take();
+            }
+        }
+
+        append_new_named(&mut self.clusters, next.clusters, |x| &x.name);
+        append_new_named(&mut self.auth_infos, next.auth_infos, |x| &x.name);
+        append_new_named(&mut self.contexts, next.contexts, |x| &x.name);
+        if self.current_context.is_none() {
+            self.current_context = next.current_context.take();
+        }
+        match self.extensions.as_mut() {
+            Some(mut base_exts) => {
+                if let Some(next_exts) = next.extensions.take() {
```

Comment:
This feels like the type of thing where `extensions` should be ` Vec<String>` with `[serde(default, skip_serialize_if = "Option::is_none")]` rather than `Option<Vec<String>>` since we treat missing as empty anyway.

Though it seems we have a convention for `Option<Vec<.>>` atm so it's a bit of a breaking change :thinking: